### PR TITLE
dlmodule can support riscv architecture

### PR DESCRIPTION
**Pull request by chunyexixiaoyu**
[ADD]riscv.c  dlmodule can support riscv architecture
[CHG]dlelf.c  support distinguishing  32-bit ELF format and 64-bit ELF format

## 拉取/合并请求描述：(PR description)

[

dlmodule可以支持risc-v 架构的芯片，本地已经验证，用户态在文件系统可以直接执行.mo 文件以及调用.so 的函数库，且.mo文件也可以调用.so的函数。该实验是在K210的板子进行实验。另外需要修改编译配置文件以及链接文件。
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [ ] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [ ] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [ ] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [ ] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [ ] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [ ] 本拉取/合并符合[RT-Thread代码规范](../documentation/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/coding_style_en.txt)

### CODE REVIEW

File: components/libc/libdl/dlelf.c [hunk 1]
```diff
@@ -38,7 +40,7 @@ rt_err_t dlmodule_load_shared_object(struct rt_dlmodule* module, void *module_pt
         if (phdr[index].p_type != PT_LOAD)
             continue;
 
-        LOG_D("LOAD segment: %d, 0x%p, 0x%08x", index, phdr[index].p_vaddr, phdr[index].p_memsz);
+        //LOG_D("LOAD segment: %d, 0x%p, 0x%08x", index, phdr[index].p_vaddr, phdr[index].p_memsz);
```
Comment (BernardXiong) [inline components/libc/libdl/dlelf.c:43]: why remove this log?

File: components/libc/libdl/dlelf.c [hunk 2]
```diff
@@ -473,3 +475,457 @@ rt_err_t dlmodule_load_relocated_object(struct rt_dlmodule* module, void *module
...
+    vstart_addr = vend_addr = RT_NULL;
+    for (index = 0; index < elf_module->e_phnum; index++)
+    {
+        if (phdr[index].p_type != PT_LOAD)
+            continue;
+
+        //LOG_D("LOAD segment: %d, 0x%p, 0x%08x", index, phdr[index].p_vaddr, phdr[index].p_memsz);
+        if (phdr[index].p_memsz < phdr[index].p_filesz)
+        {
+            rt_kprintf("invalid elf: segment %d: p_memsz: %d, p_filesz: %d\n",
```
Comment (BernardXiong) [inline components/libc/libdl/dlelf.c:506]: Please use LOG_E to replace rt_kprintf.

File: components/libc/libdl/dlelf.h [hunk 3]
```diff
@@ -91,6 +124,21 @@ typedef struct
     Elf32_Word sh_entsize;                     /* section entry size */
 } Elf32_Shdr;
 
+typedef struct
+{
+  Elf64_Word        sh_name;                /* Section name (string tbl index) */
+  Elf64_Word        sh_type;                /* Section type */
+  Elf64_Xword       sh_flags;                /* Section flags */
```
Comment (BernardXiong) [inline components/libc/libdl/dlelf.h:131]: Please keep the comments style alignment.

**Review by BernardXiong — CHANGES_REQUESTED** [review]
It's better to write code as,
```c
#ifdef _64bit
typedef rt_uint64_t Elf_Addr;
#else
typedef rt_uint32_t Elf_Addr;
#endif
```

File: components/libc/libdl/dlelf.c (hunk 2 above)
Comment (chunyexixiaoyu) [reply components/libc/libdl/dlelf.c:506]: OK,it's a historical issue and I wiil fix it

File: components/libc/libdl/dlelf.c (hunk 1 above)
Comment (chunyexixiaoyu) [reply components/libc/libdl/dlelf.c:43]: > It's better to write code as,
> 
> ```c
> #ifdef _64bit
> typedef rt_uint64_t Elf_Addr;
> #else
> typedef rt_uint32_t Elf_Addr;
> #endif
> ```
@BernardXiong  It really is a better way.

**Comment by mysterywolf** [comment]
related issue: https://github.com/RT-Thread/rt-thread/issues/4574

File: components/libc/libdl/dlelf.c [hunk 4]
```diff
@@ -18,8 +19,8 @@
 rt_err_t dlmodule_load_shared_object(struct rt_dlmodule* module, void *module_ptr)
 {
     rt_bool_t linked   = RT_FALSE;
-    rt_uint32_t index, module_size = 0;
-    Elf32_Addr vstart_addr, vend_addr;
+    rt_uintmax index, module_size = 0;
```
Comment (BernardXiong) [inline components/libc/libdl/dlelf.c:22]: Please use `rt_ubase_t`.

**Review by BernardXiong — CHANGES_REQUESTED** [review]
The `rt_base_t/rt_ubase_t` is a basic data width type for 32/64 bit system.

File: components/libc/libdl/dlelf.c (hunk 4 above)
Comment (chunyexixiaoyu) [reply components/libc/libdl/dlelf.c]: > Please use `rt_ubase_t`.
OK

**Pull request merged**